integration/kotlinx-coroutines-androidx-lifecycle/src/Lifecycle.kt

```diff
@@ -0,0 +1,92 @@
+package kotlinx.coroutines.androidx.lifecycle
+
+import androidx.lifecycle.GenericLifecycleObserver
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.Lifecycle.State.DESTROYED
+import androidx.lifecycle.Lifecycle.State.INITIALIZED
+import androidx.lifecycle.LifecycleOwner
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * Returns a [CoroutineScope] that uses [Dispatchers.Main] by default, and that is cancelled when
+ * the [Lifecycle] reaches [Lifecycle.State.DESTROYED] state.
+ *
+ * Note that this value is cached until the Lifecycle reaches the destroyed state.
+ */
+val Lifecycle.coroutineScope: CoroutineScope
```

Comment:
After your comment, I have read Supervision doc again, and now, I'm wondering if I shouldn't replace `Job()` with `SupervisorJob()`, it seems like a better default to me after reading [this](https://github.com/Kotlin/kotlinx.coroutines/blob/master/docs/exception-handling.md#supervision).
What do you think @elizarov @qwwdfsad?

Reply:
After thinking about it several times, I decided to replace `Job()` by `SupervisorJob()` in latest commit. I agree that it is a better default for this integration.
Thanks @gildor for the suggestion!

Reply:
I still not sure that it's the right solution. It should probably be explicit and a user should choose the type of Job because semantics is very different, I would discuss in this issue first
https://github.com/Kotlin/kotlinx.coroutines/issues/763